### PR TITLE
Image resource addons

### PR DESCRIPTION
**Pull request by Montellese**
This is a copy of @notspiff's work from #6917 to add image resource addons. I've only done a few minor cosmetic tweaks and added updated VS project files. Xcode sync is still missing.

The icon for the new addon type and the weather icon addon are also still missing.

### CODE REVIEW

**Comment by ronie** [comment]
if we're going to move weather icons to an addon,
it would make sense to give users an option to select a different weather icons addon.

i see it's currently hardcoded to resource.images.weather.default.
that makes little sense to me.

**Comment by uNiversaI** [comment]
defaultaddonimages
And this was the icon I didn back then for that spiff PR which got lost when it was closed

**Comment by mkortstiege** [comment]
Great addition, thanks. Once this is in @phil65 can put `script.extendedinfo` on a diet ;)

**Comment by phil65** [comment]
I will be happy to do so :) Thx @Montellese! And thx @notspiff!

**Comment by Montellese** [comment]
@ronie: I don't really know how skins handle weather icons. Can/Should every skin choose its own weather icons that match the skin's overal design or should that decision be up to the user? Depending on that it may or may not make sense to give the user the possibility to choose the weather icons themselves.

**Comment by ronie** [comment]
@Montellese some skins use the kodi provied icons, some ship with their own
some even let the user chose between the kodi ones and the one supplied by the skin.

but let's take a step back... why were the weather icons moved to an addon?
what's the purpose? there must be an idea behind it, no?

**Comment by Montellese** [comment]
@ronie: I don't know why @notspiff moved them into an addon but I guess it was to show how image resource addons could be used. And you could update the weather icons independent of any kodi release.

**Comment by MartijnKaijser** [comment]
it was a POC. I agree with ronie that users/skins need to have a choice on which "resource add-on" they will use. That was the whole concept. For example that user/skin can prefer/select having a resource that provides coloured studio logos or only alpha ones.
Then comes should they be able to pick one or several and add those to the resource path. If multiple which one will overrule the rest in case same file name exist in several addons.
Additionally i would go as far that we just don't ship this addon with Kodi and let the weather addons or skin pull it in as dependency. 
JMarshall also did some initial work some years ago to outsource the studio flags all skins use and then it was up to the user or skinner which one to use/import.
Not trying to paint anything but just making some observations.

**Comment by da-anda** [comment]
IMO weather icons could/should be handled just like UI sounds. Skins define a default to use and users can override in GUI settings.

As for generic resources - IMO skins decide, so they should be able to specify from which add-on to grab the images, maybe something like resource://media/studio-flags/foo/bar.png@[specific addon-key] ?

Another way to handle this would be that the image resource add-ons specify for which category/type the provide images for. Core then parses the categories defined in the XML and generates a settings list out of it where users can override which pack to use for what. Default would be "skin default".

File: addons/resource.language.en_gb/resources/strings.po [hunk 1]
```diff
@@ -12509,7 +12509,11 @@ msgctxt "#24034"
 msgid "Check for updates"
 msgstr ""
 
-#empty string with id 24035
+#: Name of an Add-on category
+#: xbmc/addons/Addon.cpp
+msgctxt "#24035"
+msgid "Image collections"
```
Comment (da-anda) [inline addons/resource.language.en_gb/resources/strings.po]: I'm fine with "image collections", but somehow "image packs" came to my mind when I thought about this feature.

Comment (uNiversaI) [inline addons/resource.language.en_gb/resources/strings.po]: Image packs would be better suited for general images if this ever goes beyond weather images.

**Comment by notspiff** [comment]
@da-anda read the code, the second suggestion is precisely how i designed it.

**Comment by da-anda** [comment]
@notspiff thanks for your work on this. And sorry, didn't fully understand the code in the PR as I'm not yet familiar with the underlying resource handling. So basically image resources are provided via `resource://[addon-key]/path/to/file.[whitelisted extension]`? And the specified `type=weather` can be used to filter the resource add-ons f.e. to be listed as possible select options in a setting, right? That's a very good start IMO, thanks.


### Additional thoughts
Since this is also flagged as RFC, some thoughts which could be addresssed in a follow up

In case of image resources it might be nice to not have to strictly specify an extension key in the URL but have Kodi in control to resolve a suiteable resource. So that `resource://image:weather/foo.png` would deliver form the preferred resource add-on for weather images (however we plan to resolve this in Kodi), without knowing what's installed or how the folder structure of that add-on looks like (add-ons would have to specify "entry points/folder locations" for each type they register for ofc, or we need strict conventions for folder structures in resource add-ons).
This might also help to prevent possible naming conflicts among resource add-ons, like when two add-ons have something like `media/default.png` but related to very different things (one is a studio flag, the other some actor dummy). Using this approach it would still be possible to fetch resources from a particular addon, either by using current scheme (we can detect if a "." is part of the first url segment) or by simply appending `@[add-on key]` to the URL.

Just some thoughts - nothing more.

**Comment by Montellese** [comment]
IMO in a first step we should simply get support for image packs/collections through resource addons working. Further integrating this by letting the user choose the resource addon can still be added but IMO that's not the main purpose here. The main purpose is to allow multiple skins to share the same data and the skin has full control over which resource addon is being used. After all not all image packs will fit every skin.

Otherwise we end up overcomplicating things and then the work won't be finished and we get nothing at all.

**Comment by da-anda** [comment]
see the part where I said "could be addresssed in a follow up". I only wanted to share the idea of a little different path structure/handling upfront so that early adopters won't have to adjust their skin twice in case they would prefer the other, possible follow up, approach.

**Comment by Montellese** [comment]
@uNiversaI: Thanks for the images. Integrated both.
@Memphiz: Thanks for the Xcode sync. Integrated it into the proper commit.

jenkins build this please

**Comment by uNiversaI** [comment]
@Montellese youre welcome, thx.

I only have a question regarding upgrading to this, just like the languages addons and skin UI sounds addons before I dont see any code getting rid of the old directories/images and existing files and there it started causing odd behaviours (skin UI sounds) until had to delete such things manually.
So even if the old stuff is no longer used shouldn't there be some sort of cleanup as to prevent users from having to discover and delete these remnants this manually?

**Comment by MartijnKaijser** [comment]
Good to go after a rebase?

**Comment by Montellese** [comment]
We haven't decided yet whether we actually want the weather icons as an addon or not.

**Comment by MartijnKaijser** [comment]
Well imo we do and in best case it should a dependency of weather Addons or skins. 
Regardless we really want this for other artwork packs @ronie already prepared.

File: addons/resource.language.en_gb/resources/strings.po (hunk 1 above)
Comment (Montellese) [inline addons/resource.language.en_gb/resources/strings.po]: So what's the deal here?

Comment (MartijnKaijser) [inline addons/resource.language.en_gb/resources/strings.po]: I'd be tempted to go with "image pack" as well though "image collection" works as well. Don't let that hold up the PR :)

**Comment by Montellese** [comment]
jenkins build this please

**Pull request merged**